### PR TITLE
OCR2DR - Modified AuthorizedReciever contract for beta

### PR DESCRIPTION
**Pull request by justinkaseman**
### Description
Adds an abstract contract for gating usage based on an allow list.
Unlike AuthorizedReciever.sol, this checks `tx.origin`, rather than `msg.sender`, since requests flow to the Oracle contract through the Client contract.
It is used in the OCR2DR Oracle contract for a closed beta.
Functionality of the allow list can be paused for ease of testing and to enter open beta.

### CODE REVIEW

File: contracts/src/v0.8/dev/ocr2dr/AuthorizedOriginReceiver.sol [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+  error NotAllowedToSetSenders();
+
+  EnumerableSet.AddressSet private s_authorizedSenders;
+  address[] private s_authorizedSendersList;
+
+  /**
+   * @notice Sets the fulfillment permission for a given node. Use `true` to allow, `false` to disallow.
+   * @param senders The addresses of the authorized Chainlink node
+   */
+  function setAuthorizedSenders(address[] calldata senders) external override validateAuthorizedSenderSetter {
```
Comment (bolekk) [inline contracts/src/v0.8/dev/ocr2dr/AuthorizedOriginReceiver.sol:51]: This might get inconvenient as the list grows. Why not add/remove methods?

File: contracts/src/v0.8/dev/ocr2dr/AuthorizedOriginReceiver.sol [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+    }
+    s_authorizedSendersList = senders;
+    emit AuthorizedSendersChanged(senders, msg.sender);
+  }
+
+  /**
+   * @notice Retrieve a list of authorized senders
+   * @return array of addresses
+   */
+  function getAuthorizedSenders() public view override returns (address[] memory) {
```
Comment (KuphJr) [inline contracts/src/v0.8/dev/ocr2dr/AuthorizedOriginReceiver.sol:69]: `s_authorizedSendersList` can just be set as a public variable, removing the need for a getter.

File: contracts/src/v0.8/dev/ocr2dr/AuthorizedOriginReceiver.sol [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+ * 1. Make the Oracle contract ownable, to control who can set the authorized senders
+ * ```
+ * import "../../ConfirmedOwner.sol";
+ * ...
+ * contract OCR2DROracle is OCR2DROracleInterface, OCR2Base, ConfirmedOwner {
+ * ...
+ * constructor() OCR2Base(true) ConfirmedOwner(msg.sender) {}
+ * ```
+ *
+ * 2. Extend OCR2DROracle.sol with this contract
```
Comment (bolekk) [inline contracts/src/v0.8/dev/ocr2dr/AuthorizedOriginReceiver.sol:21]: I'm guessing you'll do that in a separate PR...? Why do we need a comment here then?

**Review by bolekk — COMMENTED** [review]
two comments

File: contracts/src/v0.8/dev/ocr2dr/AuthorizedOriginReceiver.sol [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+   * @notice Sets the fulfillment permission for a given node. Use `true` to allow, `false` to disallow.
+   * @param senders The addresses of the authorized Chainlink node
+   */
+  function setAuthorizedSenders(address[] calldata senders) external override validateAuthorizedSenderSetter {
+    if (senders.length == 0) {
+      revert EmptySendersList();
+    }
+    for (uint256 i = 0; i < s_authorizedSendersList.length; i++) {
+      s_authorizedSenders.remove(s_authorizedSendersList[i]);
+    }
```
Comment (KuphJr) [inline contracts/src/v0.8/dev/ocr2dr/AuthorizedOriginReceiver.sol:57]: Do we want the `add` and `remove` functionality in a single function?  My concern is that if we have to whitelist may addresses (500+), it will require too much gas for a single tx.

File: contracts/src/v0.8/dev/ocr2dr/AuthorizedOriginReceiver.sol (hunk 3 above)
Comment (justinkaseman) [reply contracts/src/v0.8/dev/ocr2dr/AuthorizedOriginReceiver.sol:21]: If we were to do those steps now we would have to modify every tests to set up the authorized senders to the oracle.
Then once beta was over we would have to remove this and undo all of the tests.

If you think it's important to have it in the repo I can spend the time to do that.

File: contracts/src/v0.8/dev/ocr2dr/AuthorizedOriginReceiver.sol [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+    return s_authorizedSendersList;
+  }
+
+  /**
+   * @notice Use this to check if a node is authorized for fulfilling requests
+   * @param sender The address of the Chainlink node
+   * @return The authorization status of the node
+   */
+  function isAuthorizedSender(address sender) public view override returns (bool) {
+    return s_authorizedSenders.contains(sender);
```
Comment (KuphJr) [inline contracts/src/v0.8/dev/ocr2dr/AuthorizedOriginReceiver.sol:79]: This is going to require a huge amount of gas if the senders array gets too large.

We need 2 data structures to keep track of authorized senders.  A mapping & an array.  `isAuthorizedSender` should just check a mapping to determine if an address is authorized, requiring O(1) gas cost.

File: contracts/src/v0.8/dev/ocr2dr/AuthorizedOriginReceiver.sol (hunk 1 above)
Comment (justinkaseman) [reply contracts/src/v0.8/dev/ocr2dr/AuthorizedOriginReceiver.sol:51]: Good point. I can change.

File: contracts/src/v0.8/dev/ocr2dr/AuthorizedOriginReceiver.sol (hunk 4 above)
Comment (KuphJr) [reply contracts/src/v0.8/dev/ocr2dr/AuthorizedOriginReceiver.sol:57]: IMO, we should have 2 functions. `addAuthorizedSenders(address[] senders)` and `removeAuthorizedSenders(address[] senders)`

Comment (KuphJr) [reply contracts/src/v0.8/dev/ocr2dr/AuthorizedOriginReceiver.sol:57]: On second thought, even `removeAuthorizedSenders` might fail if the whitelist gets too large since we would have to loop thru every address to find which senders to remove.

We might have to implement ALL of this logic exclusively w/ a mapping, eliminating the `s_authorizedSendersList` array entirely.  This would eliminate the ability to read a list of approved addresses, but at least would never run into gas problems.

**Review by KuphJr — CHANGES_REQUESTED** [review]

File: contracts/src/v0.8/dev/ocr2dr/AuthorizedOriginReceiver.sol (hunk 4 above)
Comment (KuphJr) [reply contracts/src/v0.8/dev/ocr2dr/AuthorizedOriginReceiver.sol:57]: tbh, we shouldn't really need an array of approved addresses anyway as we can just keep this data up-to-date in a spreadsheet.
We should also add the ability to delete _every_ mapping in the array and re-add an array of authorized senders.

File: contracts/src/v0.8/dev/ocr2dr/AuthorizedOriginReceiver.sol (hunk 5 above)
Comment (justinkaseman) [reply contracts/src/v0.8/dev/ocr2dr/AuthorizedOriginReceiver.sol:79]: Contains is O(1)
https://docs.openzeppelin.com/contracts/3.x/api/utils#EnumerableSet-contains-struct-EnumerableSet-AddressSet-address-

Comment (KuphJr) [reply contracts/src/v0.8/dev/ocr2dr/AuthorizedOriginReceiver.sol:79]: Oh, sweet!

Comment (KuphJr) [reply contracts/src/v0.8/dev/ocr2dr/AuthorizedOriginReceiver.sol:79]: So does that mean we can figure out how to implement `delete` in O(1) as well to address my 1st comment?

File: contracts/src/v0.8/dev/ocr2dr/AuthorizedOriginReceiver.sol (hunk 4 above)
Comment (justinkaseman) [reply contracts/src/v0.8/dev/ocr2dr/AuthorizedOriginReceiver.sol:57]: I'll modify to add and remove because of what Bolek mentioned.
But I think that gas concerns are an edge case.
For closed beta I don't think we will have even close to that many users.
We can set expectations with Product if that is a concern.

File: contracts/src/v0.8/dev/ocr2dr/AuthorizedOriginReceiver.sol (hunk 5 above)
Comment (justinkaseman) [reply contracts/src/v0.8/dev/ocr2dr/AuthorizedOriginReceiver.sol:79]: Not if we want to be able to see the full list at once

File: contracts/src/v0.8/dev/ocr2dr/AuthorizedOriginReceiver.sol (hunk 4 above)
Comment (KuphJr) [reply contracts/src/v0.8/dev/ocr2dr/AuthorizedOriginReceiver.sol:57]: I am not sure how many beta users Bryan is expecting, but I know he was scoping out costs for a minimum of 500 users.

I would rather we support this now rather than have to modify contracts later if we run into problems.

Comment (bolekk) [reply contracts/src/v0.8/dev/ocr2dr/AuthorizedOriginReceiver.sol:70]: Let's have addAuthorizedSenders(address[] senders) and removeAuthorizedSenders(address[] senders)

File: contracts/src/v0.8/dev/ocr2dr/AuthorizedOriginReceiver.sol (hunk 3 above)
Comment (bolekk) [reply contracts/src/v0.8/dev/ocr2dr/AuthorizedOriginReceiver.sol:20]: So is the plan to always add this locally before deployment..? That's just a recipe for mistakes. And it won't be obvious what's deployed to prod.
How about we add a constructor flag "authorizeSenders" that can be switched off in tests to disable the check?

File: contracts/src/v0.8/dev/ocr2dr/AuthorizedOriginReceiver.sol (hunk 4 above)
Comment (justinkaseman) [reply contracts/src/v0.8/dev/ocr2dr/AuthorizedOriginReceiver.sol:70]: I still think this is early optimizing for what is an extreme edge case. Just a quick google search shows [this stackoverflow](https://stackoverflow.com/questions/48113615/solidity-for-loop-over-a-huge-amount-of-data) talking about 500k iterations.

The indices approach sounds fine if you are still very concerned about iterating `s_authorizedSendersList`.

**Review by bolekk — CHANGES_REQUESTED** [review]
Let's discuss the idea of not adding anything to OCR2DROracle but expecting anyone deploying the contracts to do it locally - I don't think that's reasonable. How about we add an option to enable/disable all auth checks on a global level? That way some tests can disable them to make stuff easier.

File: contracts/src/v0.8/dev/ocr2dr/AuthorizedOriginReceiver.sol [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+abstract contract AuthorizedOriginReceiver is AuthorizedReceiverInterface {
+  using EnumerableSet for EnumerableSet.AddressSet;
+
+  event AuthorizedSendersChanged(address[] senders, address changedBy);
+
+  error EmptySendersList();
+  error UnauthorizedSender();
+  error NotAllowedToSetSenders();
+
+  EnumerableSet.AddressSet private s_authorizedSenders;
```
[the remark this answers is not in the document]
Comment (justinkaseman) [reply contracts/src/v0.8/dev/ocr2dr/AuthorizedOriginReceiver.sol:43]: https://github.com/smartcontractkit/chainlink/blob/develop/contracts/STYLE.md#visibility-1

File: contracts/src/v0.8/dev/ocr2dr/AuthorizedOriginReceiver.sol (hunk 3 above)
Comment (justinkaseman) [reply contracts/src/v0.8/dev/ocr2dr/AuthorizedOriginReceiver.sol:20]: This sounds good to me. I'll make the changes today.

**Review by pinebit — DISMISSED** [review]

File: contracts/src/v0.8/dev/ocr2dr/AuthorizedOriginReceiver.sol (hunk 4 above)
Comment (KuphJr) [reply contracts/src/v0.8/dev/ocr2dr/AuthorizedOriginReceiver.sol:70]: This is probably overkill, but IMO this would be the ideal implementation for gas efficiency. (Although `getIndicesToRemove` is just for convenience and could be removed too.)

`function getIndicesToRemove(address[] calldata sendersToRemove) returns (uint256[]) external view {
    uint256[] memory indices = new uint256[](sendersToRemove.length);
    for (uint256 i = 0; i < sendersToRemove.length; i++) {
      for (uint256 j = 0; j < s_authorizedSendersList.length; j++) {
        if (sendersToRemove[i] == s_authorizedSendersList[j]) {
          indices[i] = j;
          continue;
        }
      }
    }
    return indices
  }

  /**
   * @notice Remove the permission to request for the given wallet(s).
   * @param senders The addresses of the wallet addresses to revoke access
   */
  function removeAuthorizedSenders(address[] calldata senders, uint256[] calldata indices) external validateAuthorizedSenderSetter {
    if (senders.length == 0) {
      revert EmptySendersList();
    }
    require(senders.length == indices.length, "length mismatch");
    for (uint256 i = 0; i < senders.length; i++) {
      // Check that the indices argument is correct
      if (senders[i] !== s_authorizedSendersList[indices[i]]) {
        revert IncorrectIndices();
      }
      s_authorizedSenders.remove(senders[i]);
      // Remove from s_authorizedSendersList
      address last = s_authorizedSendersList[s_authorizedSendersList.length - 1];
      s_authorizedSendersList[indices[i]] = last;
      s_authorizedSendersList.pop();
    }
    emit AuthorizedSendersChanged(senders, msg.sender);
  }`

**Review by KuphJr — APPROVED** [review]
One minor comment, but it is unnecessary so I am gonna approve.

**Comment by KuphJr** [comment]
> Let's discuss the idea of not adding anything to OCR2DROracle but expecting anyone deploying the contracts to do it locally - I don't think that's reasonable. How about we add an option to enable/disable all auth checks on a global level? That way some tests can disable them to make stuff easier.

@bolekk 

My position is, since we are deploying everything from HardHat anyway, just keep the oracle w/ whitelist in the HardHat repo.  Then we can just leave the whitelist entirely out of the contracts in the `chainlink` repo.  Since we only use HardHat to deploy, this should not lead to any mistakes.

**Comment by bolekk** [comment]
>

**Comment by bolekk** [comment]
(sorry, clicked the wrong button)

@KuphJr I don't like the idea of having different contracts in these two repos. It has just led to a mistake the other day 
 - deplyed contract was missing subIDs in events (it only happens to work because the node image is also old but will break as soon as we update it). Why not have an optional allowlist and keep the contracts identical.

**Review by KuphJr — APPROVED** [review]

**Review by pinebit — DISMISSED** [review]

**Review by bolekk — DISMISSED** [review]

**Review by bolekk — COMMENTED** [review]
One more question - I realized that you're adding AuthorizedOriginReceiver here but it doesn't exist in the hardhat repo. Why is that?

**Review by bolekk — APPROVED** [review]
